Enable detection and addition of solution level .editorconfig as solu…

mavasani opened the pull request:
…tion item

1. IDE code fix that adds new editorconfig for configuring code style/severity has been updated to create the file at root of the solution, instead of the project. Additionally, the code fix also automatically adds the file as a solution item.
2. We now detect if a new editorconfig is created at the root of the solution and provide an info bar to make it a solution item with a single click. User can choose to silence this suggestion so it never shows up again (workspace option).

Fixes #38746

sharwell commented:
:memo: Requesting items 1 and 2 above be split to separate pull requests.

mavasani commented:
Had a chat with @sharwell and we came to a compromist. Currently, there are 3 parts to this PR:
1. Change code fix to configure severity/code style that adds new editorconfig to add the newly created editorconfig file as a solution item
2. If user has an existing solution without editorconfig and they copy over or create a new editorconfig at solution root - we prompt if they want to add it as a solution item
3. Open existing solution with existing editorconfig but it is not a solution, so we prompt to add it as a solution item

3. is likely going to be intrusive for users who prefer folder view for editorconfig files as opposed to it being a solution item. I am going to revert that part to reduce the scope to scenarios where users are creating a new editorconfig.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigurationUpdater.cs:
```diff
@@ -185,9 +185,25 @@ private async Task<Solution> ConfigureAsync()
+
+            // Add the new added analyzer config document as a solution item.
+            if (PathUtilities.IsAbsolute(editorConfigDocument.FilePath) &&
+                !File.Exists(editorConfigDocument.FilePath))
```
FYI, ops like File.Exists can throw.  Is this codepath resilient to htat?

mavasani replied:
Nope, I should handle possibly exception here and also in one other place I used it in this PR.

mavasani commented:
@sharwell I have reverted 3. from https://github.com/dotnet/roslyn/pull/39127#issuecomment-539632908 and updated the PR description. PR is ready for review now.

sharwell commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AnalyzerConfigDocumentAsSolutionItemHandler.cs:
```diff
@@ -0,0 +1,130 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    [Export(typeof(AnalyzerConfigDocumentAsSolutionItemHandler)), Shared]
```
💭 I think we can omit the type if it's the same type that the attribute is applied to. When I saw the explicit type, I kept trying to figure out where it was different. 😆 

```suggestion
    [Export, Shared]
```

mavasani replied:
Thanks, fixed.

sharwell commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AnalyzerConfigDocumentAsSolutionItemHandler.cs:
```diff
@@ -0,0 +1,130 @@
+        {
+            _workspace = workspace;
+            _threadingContext = threadingContext;
+            _dte = (DTE)serviceProvider.GetService(typeof(DTE));
```
❗️ MEF constructors must not have threading dependencies (including ones from COM)

mavasani replied:
Thanks, fixed.

sharwell commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioAddSolutionItemService.cs:
```diff
@@ -0,0 +1,138 @@
+        {
+            _threadingContext = threadingContext;
+            _dte = (DTE)serviceProvider.GetService(typeof(DTE));
+            _fileChangeService = (IVsFileChangeEx)serviceProvider.GetService(typeof(SVsFileChangeEx));
```
❗️ MEF constructors must not have threading dependencies

mavasani replied:
Thanks, fixed.

JoeRobich commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigurationUpdater.cs:
```diff
@@ -185,9 +185,18 @@ private async Task<Solution> ConfigureAsync()
+                return solution;
+            }
+
+            // Add the new added analyzer config document as a solution item.
```
```suggestion
            // Add the newly added analyzer config document as a solution item.
```

JoeRobich commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AnalyzerConfigDocumentAsSolutionItemHandler.cs:
```diff
@@ -0,0 +1,136 @@
+            // Check if added analyzer config document is at the root of the current solution.
+            var analyzerConfigDocumentFilePath = e.NewSolution.GetAnalyzerConfigDocument(e.DocumentId)?.FilePath;
+            var analyzerConfigDirectory = PathUtilities.GetDirectoryName(analyzerConfigDocumentFilePath);
+            var solutionDirectory = PathUtilities.GetDirectoryName(e.NewSolution?.FilePath);
```
Why can we trust `e.NewSolution` to be non-null when calling `GetAnalyzerConfigDocument` and not when getting the `FilePath`? Is this an extension method quirk?

mavasani replied:
Hmm that is an oversight, I will fix it here.

JoeRobich commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioAddSolutionItemService.cs:
```diff
@@ -0,0 +1,155 @@
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioAddSolutionItemService(
+            IThreadingContext threadingContext,
+            SVsServiceProvider serviceProvider)
```
Is this needed since we pass an `IServiceProvider` in the `Initialize` method?

mavasani replied:
```suggestion
```

JoeRobich commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AnalyzerConfigDocumentAsSolutionItemHandler.cs:
```diff
@@ -0,0 +1,136 @@
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AnalyzerConfigDocumentAsSolutionItemHandler(
+            SVsServiceProvider serviceProvider,
```
Is this needed since we pass an `IServiceProvider` in the `Initialize` method?

mavasani replied:
```suggestion
```

mavasani commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioAddSolutionItemService.cs:
```diff
@@ -0,0 +1,155 @@
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioAddSolutionItemService(
+            IThreadingContext threadingContext,
```
```suggestion
            IThreadingContext threadingContext)
```

mavasani commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/AnalyzerConfigDocumentAsSolutionItemHandler.cs:
```diff
@@ -0,0 +1,135 @@
+            // Check if added analyzer config document is at the root of the current solution.
+            var analyzerConfigDocumentFilePath = e.NewSolution.GetAnalyzerConfigDocument(e.DocumentId)?.FilePath;
+            var analyzerConfigDirectory = PathUtilities.GetDirectoryName(analyzerConfigDocumentFilePath);
+            var solutionDirectory = PathUtilities.GetDirectoryName(e.NewSolution?.FilePath);
```
```suggestion
            var solutionDirectory = PathUtilities.GetDirectoryName(e.NewSolution.FilePath);
```

mavasani commented:
Thanks @JoeRobich
